File: android/src/main/kotlin/com/gtbluesky/fusion/container/FusionStackManager.kt
package com.gtbluesky.fusion.container

import android.app.Activity
import com.gtbluesky.fusion.Fusion
import java.lang.ref.WeakReference

internal object FusionStackManager {
    private val activityStack = mutableListOf<WeakReference<Activity>>()

    fun add(activity: Activity) {
        remove(activity)
        activityStack.add(WeakReference(activity))
    }

    fun remove(activity: Activity) {
        activityStack.removeAll {
            it.get() == activity || it.get() == null
        }
    }

    fun getTopActivity(): Activity? {
        if (activityStack.isEmpty()) return null
        return activityStack.last().get()
    }

    val containerStack = mutableListOf<WeakReference<FusionContainer>>()

    fun isEmpty() = containerStack.isEmpty()

    fun add(container: FusionContainer) {
        remove(container)
        containerStack.add(WeakReference(container))
    }

    fun remove(container: FusionContainer) {
        containerStack.removeAll {
            it.get() == container || it.get() == null
        }
    }

    fun getTopContainer(): FusionContainer? {
        if (containerStack.isEmpty()) return null
        return containerStack.last().get()
    }

    fun getTopActivityContainer(): FusionContainer? {
        if (containerStack.isEmpty()) return null
        return containerStack.findLast { it.get() is Activity }?.get()
    }

    fun findContainer(uniqueId: String): FusionContainer? {
        if (uniqueId.isEmpty()) return null
        for (containerRef in containerStack) {
            if (containerRef.get()?.uniqueId() == uniqueId) {
                return containerRef.get()
            }
        }
        return null
    }

    fun closeContainer(container: FusionContainer) {
        if (container is Activity) {
            container.finish()
            // 透明容器则关闭退出动画
            if (container.isTransparent()) {
                container.overridePendingTransition(0, 0)
            }
        }
    }

    fun isAttached(): Boolean {
        containerStack.forEach {
            if (it.get()?.isAttached() == true) {
                return true
            }
        }
        return false
    }

    fun notifyEnterForeground() {
        if (!isEmpty()) {
            Fusion.engineBinding?.engine?.lifecycleChannel?.appIsResumed()
        }
        Fusion.engineBinding?.notifyEnterForeground()
    }

    fun notifyEnterBackground() {
        if (!isEmpty()) {
            Fusion.engineBinding?.engine?.lifecycleChannel?.appIsPaused()
        }
        Fusion.engineBinding?.notifyEnterBackground()
    }
}

File: ios/Classes/Container/FusionStackManager.swift
//
//  FusionStackManager.swift
//  fusion
//
//  Created by gtbluesky on 2022/3/10.
//

import UIKit

internal class FusionStackManager {
    var containerStack = [WeakReference<FusionViewController>]()
    static let instance = FusionStackManager()

    private init() {
    }

    func add(_ container: FusionViewController) {
        remove(container)
        containerStack.append(WeakReference(container))
    }

    func remove(_ container: FusionViewController) {
        containerStack.removeAll {
            $0.value == container || $0.value == nil
        }
    }

    func getTopContainer() -> FusionViewController? {
        if containerStack.isEmpty {
            return nil
        }
        return containerStack.last?.value
    }

    func findContainer(_ uniqueId: String) -> FusionViewController? {
        if uniqueId.isEmpty {
            return nil
        }
        for containerRef in containerStack {
            if containerRef.value?.uniqueId == uniqueId {
                return containerRef.value
            }
        }
        return nil
    }

    func closeContainer(_ container: FusionViewController) {
        let vc = container
        let nc = vc.navigationController
        if let count = nc?.viewControllers.count {
            if count > 1 {
                if (nc?.topViewController == vc) {
                    nc?.popViewController(animated: true)
                } else if nc?.viewControllers.contains(vc) == true {
                    vc.removeFromParent()
                }
            } else if count == 1 && nc?.presentingViewController != nil {
                nc?.dismiss(animated: vc.isViewOpaque)
            }
        } else {
            if vc.modalPresentationStyle == .overFullScreen || vc.modalPresentationStyle == .overCurrentContext {
                var nextVc = vc.presentingViewController
                if (nextVc is UINavigationController) {
                    nextVc = (nextVc as? UINavigationController)?.topViewController
                }
                nextVc?.beginAppearanceTransition(true, animated: false)
                vc.dismiss(animated: vc.isViewOpaque) {
                    nextVc?.endAppearanceTransition()
                }
            } else {
                vc.dismiss(animated: vc.isViewOpaque)
            }
        }
    }

    func notifyEnterForeground() {
        Fusion.instance.engineBinding?.notifyEnterForeground()
    }

    func notifyEnterBackground() {
        Fusion.instance.engineBinding?.notifyEnterBackground()
    }
}
